lib/hooks/userspace/hook-grpc.js

```diff
@@ -129,6 +128,299 @@ function makeClientConstructorWrap(makeClientConstructor) {
+        if (agent.config_.enhancedDatabaseReporting) {
+          if (err) {
+            rootContext.addLabel('error', err);
+          }
+          else {
+            rootContext.addLabel('result', JSON.stringify(result));
+          }
+          if (trailer) {
+            rootContext.addLabel('trailing_metadata', JSON.stringify(trailer.getMap()));
+          }
+        }
+        if (err) {
+          // If there is an error, then no result will be serialized to
+          // be sent back to the client. So end the root span here.
+          endRootSpanForRequest(rootContext, requestName);
+        }
+        return callback(err, result);
+      };
+      return func.apply(this, arguments);
+    }, context);
+  });
+  shimmer.wrap(handlerSet, 'serialize', function (serialize) {
+    return function serializeTrace() {
+      var serializeResult = serialize.apply(this, arguments);
+      endRootSpanForRequest(rootContext, requestName);
+      return serializeResult;
+    };
+  });
+}
+
+/**
+ * Returns a function that wraps the gRPC server register function in order
+ * to create trace spans for gRPC service methods.
+ * @param {Function} register The function Server.prototype.register
+ * @returns {Function} registerTrace The new wrapper function.
+ */
+function serverRegisterWrap(register) {
+  return function registerTrace(name, handler, serialize, deserialize, method_type) {
+    // register(n, h, s, d, m) is called in addService once for each service method.
+    // Its role is to assign the serialize, deserialize, and user logic handlers
+    // for each exposed service method. Here, we wrap these functions depending on the
+    // method type.
+    var namespace = cls.getNamespace();
+    if (!namespace) {
+      agent.logger.info('gRPC: no namespace found, ignoring request');
+      return register.apply(this, arguments);
+    }
+    var result = register.apply(this, arguments);
+    var handlerSet = this.handlers[name];
+    var requestName = 'grpc:' + name;
+    // Proceed to wrap methods that are invoked when a gRPC service call is made.
+    // In every case, the function 'func' is the user-implemented handling function.
+    if (method_type === 'bidi') {
+      wrapBidi(namespace, handlerSet, requestName);
+    } else if (method_type === 'server_stream') {
+      wrapServerStream(namespace, handlerSet, requestName);
+    } else if (method_type === 'client_stream') {
+      wrapClientStream(namespace, handlerSet, requestName);
+    } else { // method_type === 'unary'
+      wrapUnary(namespace, handlerSet, requestName);
+    }
+    return result;
+  };
+}
+
+/**
+ * Creates and sets up a new root span for the given request.
+ * @param {Object} req The request being processed.
+ * @returns {!SpanData} The new initialized trace span data instance.
+ */
+function startRootSpanForRequest(req) {
+  var rootContext = agent.createRootSpanData(req, null, null, 1);
+  rootContext.span.kind = 'RPC_SERVER';
```

Comment:
My mistake. Though I think root span kind should default to `RPC_SERVER`, only being set otherwise when a custom span is created - what do you think?